src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -4139,6 +4139,101 @@ theorem sdiv_self {x : BitVec w} :
+    by_cases hy' : y.toInt = 1
+    · simp [hy', Int.ediv_one]; omega
+    · by_cases hx' : x.toInt = 0
+      · simp [hx']
+        norm_cast
+        exact Nat.two_pow_pos (w := w)
+      · have := Int.ediv_lt_self_of_pos_of_one_ne_one (x := x.toInt) (y := y.toInt) (by omega) (by omega)
+        simp; omega
+
+-- non-overflowing signed division bounds when numerator is negative, denumerator is positive
+theorem sdiv_nonpos_of_nonpos_of_nonneg {w : Nat} {x y : BitVec w} (hx : x.toInt ≤ 0) (hy : 0 ≤ y.toInt) :
+    x.toInt / y.toInt ≤ 0 := by
+  rcases w with _|w
+  · simp [of_length_zero]
+  · by_cases hx' : x.toInt = 0
+    · simp [hx']
+    · by_cases hy' : y.toInt = 0
+      · simp [hy']
+      · have := Int.ediv_neg_of_neg_of_pos (a := x.toInt) (b := y.toInt) (by omega) (by omega)
+        simp; omega
+
+-- non-overflowing signed division bounds when numerator is positive, denumerator is negative
+theorem sdiv_nonpos_of_nonneg_of_nonpos {w : Nat} {x y : BitVec w} (hx : 0 ≤ x.toInt) (hy : y.toInt ≤ 0) :
+   x.toInt / y.toInt ≤ 0 := by
+  rcases w with _|w
+  · simp [of_length_zero]
+  · by_cases hy' : y.toInt = -1
+    · simp [hy']; omega
+    · have := Int.ediv_nonpos_of_nonneg_of_nonpos (a := x.toInt) (b := y.toInt) (by omega) (by omega)
+      simp; omega
+
+theorem toInt_allOnes_sdiv_toInt_of_ne_zero {w : Nat} {y : BitVec w} (hw : 1 < w) :
+    (allOnes w).toInt / y.toInt = if y.toInt = 0 then 0 else if 0 < y.toInt then -1 else 1 := by
+  rcases w with _|_|w
+  · simp [hw]; omega
+  · simp [hw]; omega
+  · by_cases 0 < y.toInt
+    · simp [Int.sign_eq_one_of_pos (a := y.toInt) (by omega), Int.neg_one_ediv]
+      omega
+    · by_cases hy : y.toInt = 0
+      · simp [hy]
+      · simp [Int.sign_eq_neg_one_of_neg (a := y.toInt) (by omega), Int.neg_one_ediv]
+        omega
+
+-- non-overflowing signed division bounds when numerator is non positive, denumerator is less than one
+theorem sdiv_lt_of_lt_allOnes {w : Nat} {x y : BitVec w} (hx : x.toInt ≤ 0) (hy : y.toInt < - 1)  :
+    x.toInt / y.toInt < 2 ^ (w - 1) := by
+  rcases w with _|_|w
+  · simp [of_length_zero]
+  · have hy := eq_zero_or_eq_one (a := y)
+    simp [← toInt_inj, toInt_zero, toInt_one] at hy
+    omega
+  · have xle := le_two_mul_toInt (x := x); have xlt := two_mul_toInt_lt (x := x)
+    have hx' : x.toInt = 0 ∨ x.toInt = - 1 ∨ x.toInt < - 1 := by omega
+    rcases hx' with hx'|hx'|hx'
+    · simp [hx']; omega
+    · have := BitVec.toInt_allOnes_sdiv_toInt_of_ne_zero (y := y) (by omega)
+      simp only [toInt_allOnes, Nat.lt_add_left_iff_pos, Nat.zero_lt_succ, ↓reduceIte,
+        Int.reduceNeg] at this
+      simp [hx', this]
+      omega
+    · have := Int.ediv_lt_natAbs_self_of_lt_neg_one_of_lt_neg_one (x := x.toInt) (y := y.toInt) (by omega) hy
+      simp; omega
+
+-- signed division of (x y : BitVec w) is always -2 ^ w ≤ x.toInt / y.toInt
+theorem neg_two_pow_le_sdiv {x y : BitVec w} :
+    - 2 ^ (w - 1) ≤ x.toInt / y.toInt := by
```

Comment:
```suggestion
/-- For bitvectors `x, y` of width `w`, `-2 ^ w` is a lower bound of `x.toInt / y.toInt`.
theorem neg_two_pow_le_toInt_div_toInt {x y : BitVec w} :
    - 2 ^ (w - 1) ≤ x.toInt / y.toInt := by
```

I suggest a new theorem name, since the theorem, as far as I can tell, is not about `sdiv` at all, but rather a bound about `x.toInt / y.toInt`.

Reply:
I see. In this case, I would also propose changing the def of `sdivOverflow` to use `sdiv`, though, and update everything accordingly. 
In particular, I would expect to unfold the `sdiv` operation (`BitVec.sdiv_eq`) and match based on the `msb` (instead of the signs of `toInt`) and _then_ refer to the corresponding `Int.*` theorems. I am not sure which of these strategies is better, suggestions are welcome.

Reply:
I fundamentally see two ways in which we can prove this set of `sdivOverflow`-related theorems: 
1. base these theorems on `sdiv` (e.g. `x.sdiv y`) and based on the `msb` of `x` and `y` reference different `ediv` theorems 
2. use `toInt` and directly refer `Int.ediv` operation reasoning about the signs of `toInt` (current strategy)
I'm also happy to switch to (1) and open a new PR to anyways upstream the theorems I wrote using `toInt` (I believe they are anyways useful). @tobiasgrosser, I'd be interested to know what you think of this!

Reply:
I have now updated the names of the theorems using `toInt`, but will soon open a new PR with the corresponding `sdiv` statements.